### ISSUE TITLE
Initial test file request not hitting most hooks

### ISSUE THREAD

**Issue by FredKSchott** [modernweb-dev/web#594]
Progress on the Snowpack test runner continues! I almost have something working, but ran into issues with the plugin hooks interface. I need to transform the initial test file request, since files are not hosted 1:1 at their source location on disk. Create Snowpack App, for example, promotes serving your `src/` directory out of the `/_dist/*` URL. Requesting `/src/App.test.jsx`, for example, fails.

I would expect the initial test file to hit some hook where I could tell @web/test-runner that `/src/App.test.jsx` should really be tested by hitting `/_dist/App.test.js` (where the file is expected to exist).

No hooks seem to exist to catch the initial test file request (only follow up imports) and transform / redirect it to the correct location. Is this possible?

**Comment by LarsDenBakker** [comment]
If you look at this example: https://github.com/modernweb-dev/example-projects/blob/serve-test-file-example/preact-jsx/web-test-runner.config.mjs

Running the tests gives me:

```
transform /
resolveImport /
transform /__web-test-runner__/test-framework/Users/ri04pi/Workspace/code/mine/web-example-projects/preact-jsx/node_modules/@web/test-runner-mocha/dist/autorun.js
serve /test/app.test.jsx
transform /test/app.test.jsx
resolveImport /test/app.test.jsx
resolveImport /test/app.test.jsx
resolveImport /test/app.test.jsx
resolveImport /test/app.test.jsx
serve /node_modules/preact/dist/preact.module.js
serve /node_modules/@esm-bundle/chai/esm/chai.js
serve /src/App.jsx
serve /test/test-helpers.jsx
transform /node_modules/preact/dist/preact.module.js
transform /node_modules/@esm-bundle/chai/esm/chai.js
transform /src/App.jsx
transform /test/test-helpers.jsx
resolveImport /src/App.jsx
resolveImport /test/test-helpers.jsx
```

This example uses a root level `test` directory, but this could be `serve` too. You can use this hook to serve from `/_dist/` instead, or is that not what you're looking for?

**Comment by FredKSchott** [comment]
Yea, I can clarify. I'm looking for something that I can hook into BEFORE `serve /test/app.test.jsx`, so that I would see the following (notice the final URL change):

```
transform /
resolveImport /
transform /__web-test-runner__/...
<<TODO: HOOK>> /src/app.test.jsx
serve /_dist_/app.test.js
```

The request is to be able to modify the request made for each test file to point to the correct Snowpack-hosted location.

I'll play around with moving tests out of src/ and into a new test/ dir to see if this could also solve our problem, but supporting test files alongside source files (`App.jsx` + `App.test.jsx` in the same repo) would be required for larger Snowpack support.

**Comment by LarsDenBakker** [comment]
The way the initial test file is imported right now it's not passed through the resolveImport / transformImport hook. I ran into a use case for another plugin as well, I'll look into adding this.

**Comment by FredKSchott** [comment]
That would be great, thank you!

**Comment by FredKSchott** [comment]
One thing that will come up: is there a way to detect when an import is one of these initial request / test file URLs? In our plugin, we would want to only transform these test file URLs, and no others. 

Another solution would be a way for the plugin to see the initial glob that web test runner received (`ex:  web-test-runner 
 "src/**/*.test.jsx"`) and then match everything in `resolveImport` against that glob to check if it matches.

**Comment by FredKSchott** [comment]
Anything that I can do to help here? If it's easy enough for a beginner, I'm happy to take a look!

**Comment by FredKSchott** [comment]
Anything that I can do to help here? If it's easy enough for a beginner, I'm happy to take a look! Excited to move away from Jest for Snowpack apps :)

**Comment by LarsDenBakker** [comment]
Thanks for the reminder, I was a bit zoned into fixing some issues with the saucelabs launcher. I'm very excited about the move as well!

The implementation is a bit deep in the core of the test runner, I went ahead and did it in https://github.com/modernweb-dev/web/pull/634

Because the plugin interface is actually from web dev server, I don't have a great way right now to do anything custom there like passing an array of test files. Especially to make it work typesafe. That's something I need to work on. 

A test file import has a special query parameter, we can recognize it using that. I added a helper function for that, so that you don't need to hardcode it.

This should do the trick:

```js
import { isTestFilePath } from '@web/test-runner';

export default {
  plugins: [
    {
      transformImport({ source }) {
        if (isTestFilePath(source)) {
          console.log('transforming', source);
          // quick and dirty
          return source.replace('src', '_dist');
        }
      },
    },
  ],
};
```
